Expected exception rewriting

dkudernatsch opened the pull request.

dkudernatsch committed:
Project directory setup

dkudernatsch committed:
+ Treat all warning as errors
+ Enable nullable reference types for other projects

dkudernatsch committed:
Updated .gitignore to exclude entire .idea folder
removed .idea from version control

dkudernatsch committed:
Initial prototype for ExpectedException rewriting

dkudernatsch committed:
Fixed stuff

dkudernatsch committed:
Updated Test solution to mimic real life project
* added project references

dkudernatsch committed:
Changed project from .net core 3.1 to .net framework 4.7.2

Due to the way MsBuildWorkspace works internally, the .net core version of the library is unable to load and resolve project references of projects targeting framework.

dkudernatsch committed:
Removed nullable references due to bad support in .net framework

dkudernatsch committed:
Test solution unit test now run correctly

dkudernatsch committed:
Added Diagnostics reporting when opening, loading and compiling a solution

dkudernatsch committed:
Added unit tests for creation of expectedException models

dkudernatsch committed:
added test for attribute removal

dkudernatsch committed:
removed old sample test file

dkudernatsch committed:
implemented correct formatting for generated code

dkudernatsch committed:
Refactored unittests added custom nunit constraint for syntax nodes
reformatted testsolution

dkudernatsch committed:
Support for nunit well known exceptions
finished formatting

dkudernatsch committed:
reformatted and rearragned code

dkudernatsch committed:
added missing file headers

dkudernatsch committed:
Merge branch 'master' into ExpectedExceptionRewriting

Stefan-Ilic commented on ConverterConsole/ConverterConsole.csproj:
```diff
@@ -1,18 +1,20 @@
+        <OutputType>Exe</OutputType>
+        <TargetFramework>net472</TargetFramework>
+        <RootNamespace>NUnit2To3SyntaxConverter.ConverterConsole</RootNamespace>
+        <LangVersion>8</LangVersion>
```
I believe you forgot to enable nullable reference types in all projects?

Stefan-Ilic commented on ConverterConsole/CmdLineOptions.cs:
```diff
@@ -0,0 +1,41 @@
+{
+  public class CmdLineOptions
+  {
+    public CmdLineOptions (string solutionPath, string msBuildVersion = null, string msBuildPath = null)
```
I believe we should add the `Jetbrains.Annotations` package to this project and add the `UsedImplicitly` attribute to this class

Stefan-Ilic commented on ConverterConsole/Program.cs:
```diff
@@ -11,17 +12,108 @@
+              _ => Task.FromResult (1));
+    }
+
+    public static async Task<int> RunConverter (CmdLineOptions options)
```
Can be made private

Stefan-Ilic commented on ConverterConsole/Program.cs:
```diff
@@ -11,17 +12,108 @@
+
+    private static VisualStudioInstance LocateMsBuild (CmdLineOptions options)
+    {
+      // ensure a user supplied msbuild is always found
```
This method seems very long and has quite a few comments. Please try splitting it into smaller methods

Stefan-Ilic replied:
That should also eliminate the need for comments in the first place

Stefan-Ilic commented on ConverterConsole/Program.cs:
```diff
@@ -11,17 +12,108 @@
+      return queriedInstances.SingleOrDefault();
+    }
+
+    private class ConsoleProgressReporter : IProgress<ProjectLoadProgress>
```
I don't think I told you this so it's not your fault but we usually place nested types as the first "member" of a class, meaning at the very top

Stefan-Ilic commented on Core/NUnitMigrator.cs:
```diff
@@ -0,0 +1,82 @@
+    {
+      foreach (var project in solution.Projects)
+      {
+        var comp = await project.GetCompilationAsync();
```
Why aren't you using the compilation result?

dkudernatsch replied:
this loop is used to precache the compilations.
should i let roslyn compile on demand or use a discard here?

Stefan-Ilic replied:
If there is any merit in doing the compilations now vs later, not storing the result is probably fine

iam-mholle replied:
Just out of curiosity, do you need to compile them one by one or could you do the following?
```csharp
await Task.WhenAll (solution.Projects.Select (project => project.GetCompilationAsync()));
```

dkudernatsch commented on Core/ISyntaxTransformer.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace NUnit2To3SyntaxConverter
+{
+  public interface ISyntaxTransformer<TSyntax, in TContext>
```
ISyntaxTransformer needs context for some transformations e.g. for ExpectedExceptionAttributeRemover 
but for other it doesn't this is incompatible with definition

Stefan-Ilic commented on Core/NUnitMigrator.cs:
```diff
@@ -0,0 +1,82 @@
+      }
+
+      var tasks = solution.Projects
+          .Where (project => _options.ProjectFilter.Invoke (project))
```
You can remove the expression syntax and pass the `Func directly`, looks a bit nicer

Stefan-Ilic replied:
Also 2 lines below

Stefan-Ilic commented on Core/NUnitMigrator.cs:
```diff
@@ -0,0 +1,82 @@
+          .Where (project => _options.ProjectFilter.Invoke (project))
+          .SelectMany (project => project.Documents)
+          .Where (document => _options.SourceFileFilter.Invoke (document))
+          .Select (async document => (Original: document, New: await ConvertDocument (document, new ExpectedExceptionDocumentConverter())))
```
There's a lot happening here and it's a little bit hard to understand for me. Maybe private methods for the operations inside the two `Selects`?

Stefan-Ilic commented on Core/NUnitMigrator.cs:
```diff
@@ -0,0 +1,82 @@
+      await Task.WhenAll (tasks);
+    }
+
+    public async Task<Document> ConvertDocument (Document document, params IDocumentConverter[] converters)
```
Private

Stefan-Ilic commented on Core/NUnitMigrator.cs:
```diff
@@ -0,0 +1,82 @@
+      return document.WithSyntaxRoot (await newDoc.GetSyntaxRootAsync());
+    }
+
+    public async Task WriteBack (Document original, Document newDocument)
```
Private

Stefan-Ilic commented on Core/NUnitMigrator.cs:
```diff
@@ -0,0 +1,82 @@
+
+    public async Task WriteBack (Document original, Document newDocument)
+    {
+      var oldRootNode = await original.GetSyntaxRootAsync();
```
Rename `oldRootNode` to `originalRootNode`? Just a tiny thing but keeping things consistent can be very challenging in the long run

Stefan-Ilic commented on Core/NUnitMigrator.cs:
```diff
@@ -0,0 +1,82 @@
+      }
+      catch (Exception ex)
+      {
+        throw new InvalidOperationException (string.Format ("Unable to write source file '{0}'.", newDocument.FilePath), ex);
```
Why not use string interpolation?

Stefan-Ilic commented on Core/NUnitMigrator.cs:
```diff
@@ -0,0 +1,82 @@
+
+        newRootNode!.WriteTo (writer);
+      }
+      catch (Exception ex)
```
Are there more specific exception types you're expecting here that you could catch instead?

Stefan-Ilic commented on Core/MigrationOptions.cs:
```diff
@@ -0,0 +1,36 @@
+{
+  public class MigrationOptions
+  {
+    private static readonly Encoding DefaultEncoding = Encoding.Default;
```
StyleGuide says this should be `s_defaultEncoding`

Stefan-Ilic commented on Core/MigrationOptions.cs:
```diff
@@ -0,0 +1,36 @@
+
+    public static MigrationOptions DefaultOptions => new MigrationOptions();
+
+    public Func<Project, bool> ProjectFilter { get; set; } = _ => true;
```
This feels like it should be immutable. You should probably remove the setters, offer a ctor that sets the methods and keep an empty ctor that initializes with default values. This way "DefaultOptions" can always be the same object without fear of mutations

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionMethodBodyTransformer.cs:
```diff
@@ -0,0 +1,97 @@
+{
+  public class ExpectedExceptionMethodBodyTransformer : ISyntaxTransformer<MethodDeclarationSyntax, ExpectedExceptionModel>
+  {
+    public MethodDeclarationSyntax Transform (MethodDeclarationSyntax node, ExpectedExceptionModel model)
```
Why not use the interface type `IExpectedExceptionModel `?

Stefan-Ilic commented on Core/ISyntaxTransformer.cs:
```diff
@@ -0,0 +1,28 @@
+  public interface ISyntaxTransformer<TSyntax, in TContext>
+      where TSyntax : SyntaxNode
+  {
+    TSyntax Transform (TSyntax node, TContext context);
```
Perhaps context needs a better name, perhaps it's already perfect. Maybe we'll see when we have more Transformers

Stefan-Ilic commented on Core/Library.cs:
```diff
@@ -1,4 +1,5 @@
 ﻿#region copyright
```
Delete this file pls

Stefan-Ilic commented on Core/NUnitAssemblyFilter.cs:
```diff
@@ -0,0 +1,44 @@
+
+namespace NUnit2To3SyntaxConverter
+{
+  public class NUnitAssemblyFilter
```
Am I missing something or is this class never used? Both `ExpectedExceptionRewriter` and `ExpectedExceptionRewriter` initialize it but then don't use it

Stefan-Ilic commented on UnitTests/CompiledSourceFileProvider.cs:
```diff
@@ -0,0 +1,59 @@
+    public static (SemanticModel, SyntaxNode) LoadSemanticModel (string file)
+    {
+      var src = File.ReadAllText (TestContext.CurrentContext.TestDirectory + "/" + file);
+
```
Remove one empty line

Stefan-Ilic commented on UnitTests/CompiledSourceFileProvider.cs:
```diff
@@ -0,0 +1,59 @@
+
+namespace NUnit2To3SyntaxConverter.UnitTests
+{
+  public class CompiledSourceFileProvider
```
You can go ahead and make this static

Stefan-Ilic replied:
Have you thought about moving this to a `Helpers` folder?

dkudernatsch replied:
should `NUnitExtension` and `WhiteSpaceIgnoringSyntaxComparer` also go into the helpers folder?

Stefan-Ilic replied:
Hmm, I'm not sure. To me, `CompiledSourceFileProvider` seems like a "Implementation Detail" of the test, while the others are probably more "CustomAsserts" or something. What do you think?

Stefan-Ilic commented on Core/ConversionUnit.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace NUnit2To3SyntaxConverter
+{
+  public class ConversionUnit
```
This class is currently not being used. Do we still need it?

Stefan-Ilic commented on UnitTests/resources/ExpectedExceptionFromAttributeTests.cs:
```diff
@@ -0,0 +1,76 @@
+
+namespace UnitTests
+{
+    public class ExpectedExceptionFromAttributeTests
```
I'm missing an easier 1:1 mapping from tests to test cases, e.g. I think this should be called `ParseExpectedExceptionModelTestCases`

Stefan-Ilic replied:
Actually, this can be applied to the test class name as well. I believe `ParseExpectedExceptionModelTest` should be `ExpectedExceptionModelTest`

Stefan-Ilic commented on UnitTests/ExpectedException/ParseExpectedExceptionModelTest.cs:
```diff
@@ -0,0 +1,162 @@
+{
+  public class ParseExpectedExceptionModelTest
+  {
+    private AttributeData LoadAttribute (string fileName, string methodName)
```
Please move private helper methods in tests always to the bottom of the file

Stefan-Ilic replied:
Also probably rename to `LoadAttributeData`

Stefan-Ilic commented on UnitTests/ExpectedException/ParseExpectedExceptionModelTest.cs:
```diff
@@ -0,0 +1,162 @@
+
+    [Test]
+    [TestCase ("resources/ExpectedExceptionFromAttributeTests.cs", "SimpleBaseCase")]
+    public async Task NoArgsAttribute (string fileName, string methodName)
```
Once the test class is renamed, the tests should also be renamed to `[NameOfTestedMethod]_[ConcreteCase]` e.g. `CreateFromAttributeData_NoArgsAttribute` etc

Stefan-Ilic commented on UnitTests/ExpectedException/ParseExpectedExceptionModelTest.cs:
```diff
@@ -0,0 +1,162 @@
+      Assert.That (expectedExceptionModel, Is.Not.Null);
+      Assert.That (
+          expectedExceptionModel.ExceptionType,
+          Is.EquivalentTo (TypeOfExpression (IdentifierName ("Exception"))));
```
Did resharper break these lines for you? Many of these don't exceed our 170 characters per line limit and can be written in one line

Stefan-Ilic commented on UnitTests/ExpectedException/ParseExpectedExceptionModelTest.cs:
```diff
@@ -0,0 +1,162 @@
+
+namespace NUnit2To3SyntaxConverter.UnitTests.ExpectedException
+{
+  public class ParseExpectedExceptionModelTest
```
Was it a deliberate choice to not unit test `AsConstraintExpression` and `WithMessage` of `ExpectedExceptionmodel`? Since these are the unit tests for it

Stefan-Ilic commented on UnitTests/resources/ExpectedExceptionFromAttributeTests.cs:
```diff
@@ -0,0 +1,76 @@
+        {
+        }
+
+        [ExpectedException(typeof(DivideByZeroException)]
```
You're missing a closing parenthesis here. It doesn't seem to bother it though

Stefan-Ilic commented on UnitTests/ExpectedException/ParseExpectedExceptionModelTest.cs:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    [Test]
+    [TestCase ("resources/ExpectedExceptionFromAttributeTests.cs", "SimpleBaseCase")]
```
Since the file seem to be the same for all tests in this file, we should probably put it in a const string member  at the very top of the testclass and use that member in `LoadAttribute` to remove clutter from the separate tests

Stefan-Ilic replied:
Actually, instead of referencing the resources, could we just use input strings inside the tests like `var attribute = "[ExpectedException (typeof(DivideByZeroException), ExpectedExceptionName = "DivideByZeroException")]"` everywhere? It doesn't seem to add too much clutter in this test class

(We may have talked about this, but perhaps your `CompiledSourceFileProvider` could dinamycally add a method body below the specified attribute since it's always an empty method in these testcases?)

Stefan-Ilic commented on Core/ExpectedException/IExpectedExceptionModel.cs:
```diff
@@ -0,0 +1,34 @@
+    ExpressionSyntax MatchType { get; }
+    ExpressionSyntax ExpectedMessage { get; }
+    Task<AttributeSyntax> GetAttributeSyntax ();
+
```
This newline should probably be moved up to separate properties from methods?

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionModel.cs:
```diff
@@ -0,0 +1,151 @@
+    }
+
+
+    public ExpressionSyntax ExceptionType { get; }
```
Please moeve these above the ctor, also please remove the double newlines above

Stefan-Ilic commented on Core/ExpectedException/WellKnownExceptions.cs:
```diff
@@ -0,0 +1,54 @@
+{
+  public static class WellKnownExceptions
+  {
+    public static ExpressionSyntax ThrowsExceptionConstrainSyntax (string exceptionName)
```
Is this more like a `GetThrowsExceptionConstrainSyntax ` or `CreateThrowsExceptionConstrainSyntax `?

Stefan-Ilic commented on Core/ExpectedException/WellKnownExceptions.cs:
```diff
@@ -0,0 +1,54 @@
+    {
+      var exceptionNameWithoutSystemNs = Regex.Replace (exceptionName, "^(System.)", "");
+      return IsWellKnownException (exceptionNameWithoutSystemNs)
+          ? (ExpressionSyntax) MemberAccess (IdentifierName ("Throws"), exceptionNameWithoutSystemNs)
```
We usually use the ternary conditional operator only if we have simple outcomes like fixed values or values stored in variables. For things like this it's probably better to use `if` statements

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionModel.cs:
```diff
@@ -0,0 +1,151 @@
+    public static ExpectedExceptionModel CreateFromAttributeData (AttributeData attribute)
+    {
+      var attributeSyntax = attribute.ApplicationSyntaxReference.GetSyntax() as AttributeSyntax;
+      Debug.Assert (attributeSyntax != null);
```
What happens if attributeSyntax is null in a `Release` configured build?

Stefan-Ilic replied:
Asking this question basically for every occurrence of `Debug`

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionModel.cs:
```diff
@@ -0,0 +1,151 @@
+          : WithArgsExpectedExceptionModel (attribute, attributeArguments.Value);
+    }
+
+    private static ExpectedExceptionModel WithArgsExpectedExceptionModel (
```
Methods should usually be verbs, in this case they should probably be prefixed with `Create` even if they're private

Stefan-Ilic commented on Core/Extensions/SyntaxFactoryExtensions.cs:
```diff
@@ -0,0 +1,55 @@
+
+namespace NUnit2To3SyntaxConverter.Extensions
+{
+  public static class SyntaxFactoryUtils
```
File name and type name do not match

Stefan-Ilic commented on Core/Extensions/SyntaxFactoryExtensions.cs:
```diff
@@ -0,0 +1,55 @@
+      return InvocationExpression (expression, ArgumentList (argList).WithLeadingTrivia (methodCallWhiteSpace));
+    }
+
+    public static MemberAccessExpressionSyntax MemberAccess (ExpressionSyntax expression, params string[] accessor)
```
Again, it's a little confusing to see Methods that aren't named after verbs. Usually `GetX` or `CreateX` will do the trick

Stefan-Ilic replied:
I mean that for every method in this class and other classes

dkudernatsch replied:
The methods in this file are specifically named like this to conform with the naming of SyntaxFactory

Stefan-Ilic commented on Core/Extensions/SyntaxFactoryExtensions.cs:
```diff
@@ -0,0 +1,55 @@
+      var methodCallWhiteSpace = args.Count > 0
+          ? Whitespace (" ")
+          : Whitespace ("");
+      var argList = SeparatedList (args.Select (args => Argument (args).WithLeadingTrivia()));
```
Please rename one of the args

Stefan-Ilic commented on Core/ExpectedException/WellKnownExceptions.cs:
```diff
@@ -0,0 +1,54 @@
+
+namespace NUnit2To3SyntaxConverter.ExpectedException
+{
+  public static class WellKnownExceptions
```
Judging by what it's doing it seems more like a `ThrowsExpressionSyntaxFactory` or something along those lines?

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionModelBuilder.cs:
```diff
@@ -0,0 +1,98 @@
+          MatchType);
+    }
+
+    public ExpectedExceptionModelBuilder WithExceptionTypeOrName (ExpressionSyntax value)
```
This method should probably be above `Build` for consistency

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionModelBuilder.cs:
```diff
@@ -0,0 +1,98 @@
+        MatchType = MemberAccess (IdentifierName ("MessageMatch"), "Exact");
+
+      if (ExceptionName != null && ExceptionType != null)
+        throw new ArgumentException (
```
I believe this is more of an `InvalidOperationException`

Stefan-Ilic commented on UnitTests/ExpectedException/RemoveExpectedAttributeTest.cs:
```diff
@@ -0,0 +1,53 @@
+  public class RemoveExpectedAttributeTest
+  {
+    [Test]
+    [TestCase ("resources/RewriteMethodBodyTests.cs", "MultipleAttributesInAttributeListFirst")]
```
Same questions as in `ParseExpectedExceptionModelTest`: Can we only specify the attribute as strings and not use the resource file by "generating" an empty method around the attribute? And if not, we should probably store the resource file name in a constant and remove clutter from the testcases

Stefan-Ilic replied:
And if we can't go without the resource files, resource file name and test file name need to match

Stefan-Ilic commented on UnitTests/ExpectedException/RemoveExpectedAttributeTest.cs:
```diff
@@ -0,0 +1,53 @@
+    [TestCase ("resources/RewriteMethodBodyTests.cs", "MultipleAttributesInAttributeListLast")]
+    public void RemovesSingleAttribute (string fileName, string methodName)
+    {
+      var (methodSymbol, methodSyntax) = CompiledSourceFileProvider.LoadMethod (fileName, methodName);
```
You can go ahead and discard the unused `methoxdSymbol`

Stefan-Ilic commented on UnitTests/ExpectedException/FeatureIntegrationTest.cs:
```diff
@@ -0,0 +1,77 @@
+{
+  public class FeatureIntegrationTest
+  {
+    public (ExpectedExceptionRewriter, MethodDeclarationSyntax) RewriterFor (string file, string methodName)
```
Make private and move to bottom

Stefan-Ilic replied:
Also probably rename to `CreateRewriterFor`

Stefan-Ilic commented on UnitTests/ExpectedException/FeatureIntegrationTest.cs:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    [Test]
+    [TestCase (
```
For these tests you probably do really need the resource files, but you can still remove clutter by storing them in constants and only passing the method name in the test case

Stefan-Ilic commented on UnitTests/ExpectedException/FeatureIntegrationTest.cs:
```diff
@@ -0,0 +1,77 @@
+        "/resources/ExpectedExceptionIntegrationTestCases.cs",
+        "/resources/ExpectedExceptionIntegrationTestCasesExpected.cs",
+        "WithMultiLineString")]
+    public void FormatSimpleCase (string filename, string filenameExpected, string method)
```
Is this testing the simple cases or all cases?

Stefan-Ilic commented on UnitTests/resources/ExpectedExceptionIntegrationTestCases.cs:
```diff
@@ -0,0 +1,75 @@
+      Console.WriteLine();
+    }
+    
+    [ExpectedException (typeof (ArgumentNullException)]
```
How come he doesn't care about the missing closing parenthesis?

dkudernatsch replied:
because roslyn is too smart somtimes

Stefan-Ilic commented on UnitTests/resources/ExpectedExceptionIntegrationTestCasesExpected.cs:
```diff
@@ -0,0 +1,81 @@
+
+namespace UnitTests
+{
+  public class ExpectedExceptionIntegrationTests
```
I know this is not the usual way things are done, but what if the pre-transformation methods and post-transformatin methods (aka these) are in the same file, meaining very close to each other. This way we could see the difference at one glance. The input/expected pairs could be separated by a comment of "============". Just a thought, not sure if that's even good

dkudernatsch replied:
to do this would require a lot of logic in the test for loading and comparing the methods

Stefan-Ilic commented on UnitTests/resources/TestSolution/UnitTests/packages.config:
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+
+<packages>
+    <package id="NUnit" version="2.6.2" targetFramework="net472" />
```
Could you add the version number to the name of `nunit.framework.dll` since it seems kind of important and essential

Stefan-Ilic commented on UnitTests/resources/TestSolution/TestSolution.sln:
```diff
@@ -0,0 +1,34 @@
+﻿
```
Does the TestSolution serve any purpose now? It doesn't appear to be used by any test

Stefan-Ilic commented on UnitTests/resources/ExpectedExceptionBaseCase.txt:
```diff
@@ -0,0 +1,17 @@
+﻿using System;
```
Is this file used by anything?

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionRewriter.cs:
```diff
@@ -0,0 +1,67 @@
+{
+  public class ExpectedExceptionRewriter : CSharpSyntaxRewriter
+  {
+    private readonly NUnitAssemblyFilter _assemblyFilter;
```
Unused field

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionRewriter.cs:
```diff
@@ -0,0 +1,67 @@
+  public class ExpectedExceptionRewriter : CSharpSyntaxRewriter
+  {
+    private readonly NUnitAssemblyFilter _assemblyFilter;
+    private readonly ISyntaxTransformer<MethodDeclarationSyntax, ExpectedExceptionModel> _attributeRemover;
```
Could you use the interface type `IExpectedExceptionModel` in this file (except in return types)?

Stefan-Ilic replied:
Actually, even though it's probably irrelevant, as a thought experiment: The way this class is currently written, we probably should't use `ISyntaxTransformer<MethodDeclarationSyntax, ExpectedExceptionModel>` as the types of the fields, since not any `ISyntaxTransformer<MethodDeclarationSyntax, ExpectedExceptionModel>` would do because we're  using them in a very specific way. 

For true loose coupling, we'd need to store an IEnumerable of `ISyntaxTransformer<MethodDeclarationSyntax, ExpectedExceptionModel>` and apply their transformations one by one, and since that wouldn't make sense here it's perhaps better to store the more concrete type of the the `ISyntaxTransformer`s

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionRewriter.cs:
```diff
@@ -0,0 +1,67 @@
+      var methodSymbol = _semanticModel.GetDeclaredSymbol (node);
+      var attributes = methodSymbol.GetAttributes();
+      return attributes
+          // .Where (attribute => _assemblyFilter.IsSupportedAssembly (attribute.AttributeClass.ContainingAssembly.Identity))
```
Undeleted comment

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionMethodBodyTransformer.cs:
```diff
@@ -0,0 +1,97 @@
+                  ExpressionStatement (assertThatThrows).WithTrailingTrivia (Whitespace (Formatting.NewLine)))));
+    }
+
+    private static LambdaExpressionSyntax MakeLambdaExpression (StatementSyntax bodyStatement)
```
We need to pay attention to "one word per concept". Hence we probably shouldn't mix `Create`, `Get`, `Generate`, `Make`

Stefan-Ilic commented on Core/Extensions/SyntaxNodeExtension.cs:
```diff
@@ -0,0 +1,88 @@
+    private class IndentationNodeVisitor : CSharpSyntaxRewriter
+    {
+      private readonly string _indentation;
+      private readonly string _newLine;
```
Unused field

Stefan-Ilic reviewed (changes requested):
Phew, quite a few files. It probably makes sense to read all the comments before you start working on the requested changes, since I go back and forth on some issues before coming to a clear conclusion. Bear in mind that many of these are my opinions and are up for discussion if you disagree with them

dkudernatsch committed:
Added support for NullableReferenceTypes

dkudernatsch committed:
added Jetbrains annotations to mark CmdLineOption as UsedImplicitly

dkudernatsch committed:
cleaned up LocateMsbuild method

dkudernatsch committed:
made MigrationOptions immutable

dkudernatsch committed:
review changes for NUnitMigrator

dkudernatsch committed:
use interface instead of implementation

dkudernatsch committed:
removed unused classes

dkudernatsch committed:
extracted test file name to private const
renamed file to match tested file

dkudernatsch committed:
moved testcases form separate resource file to inline declaration

dkudernatsch committed:
renamed stuff

dkudernatsch committed:
changed formatting

dkudernatsch committed:
made private things private

dkudernatsch committed:
added version number to bundled nunit assembly

dkudernatsch committed:
fixed for pr review

dkudernatsch committed:
fixed file names in tests

dkudernatsch committed:
moved private class to conform with styleguide

dkudernatsch committed:
cleaned up Migrate

dkudernatsch committed:
moved compiledSourceFileProvider to Helpers folder
added tests for AsConstraintExpression

dkudernatsch committed:
finished resolving pr review

dkudernatsch committed:
finished resolving pr review

Stefan-Ilic commented on ConverterConsole/Program.cs:
```diff
@@ -50,7 +63,26 @@ public static async Task<int> RunConverter (CmdLineOptions options)
+    {
+      var queriedInstances = QueryVisualStudioInstances (options).ToList();
+
+      if (queriedInstances.Count <= 0)
```
This can never be <0, correct? So  `queriedInstances.Count == 0` is probably more idiomatic in almost all cases dealing with collections

Stefan-Ilic commented on ConverterConsole/Program.cs:
```diff
@@ -50,7 +63,26 @@ public static async Task<int> RunConverter (CmdLineOptions options)
+        return null;
+      }
+
+      return queriedInstances.SingleOrDefault();
```
Can this ever be null?

Stefan-Ilic commented on ConverterConsole/Program.cs:
```diff
@@ -65,54 +97,33 @@ private static VisualStudioInstance LocateMsBuild (CmdLineOptions options)
       // filter by supplied msbuild file
```
Are these comments still necessary? If yes perhaps it needs to be broken down even further? If no you can probably just delete them

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionAttributeRemover.cs:
```diff
@@ -28,9 +28,9 @@ public class ExpectedExceptionAttributeRemover : ISyntaxTransformer<MethodDeclar
-
+      var toRemove = model.GetAttributeSyntax().GetAwaiter().GetResult()
+          ?? throw new ArgumentException ($"Unable to get syntax tree of syntax node {nameof(node)}.\r\nNode: ({node.ToString()})");
+      
```
Be wary of "empty" lines" with spaces in them. These tend to stick around for YEARS because everyone is too lazy to create an extra commit to fix existing formatting errors

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionDocumentConverter.cs:
```diff
@@ -26,12 +26,11 @@ public class ExpectedExceptionDocumentConverter : IDocumentConverter
-      Debug.Assert (syntaxRoot != null);
-
+      var semanticModel = await document.GetSemanticModelAsync()
+          ?? throw new ArgumentException("Document does not support a semantic model.");
```
The `using System.Diagnostics` directive is not necessary anymore and should be removed everywhere

Stefan-Ilic replied:
Also, the formatting is off in this file

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionModel.cs:
```diff
@@ -44,16 +49,10 @@ public class ExpectedExceptionModel : IExpectedExceptionModel
     {
-      return await _attributeData.ApplicationSyntaxReference.GetSyntaxAsync() as AttributeSyntax;
+      return await _attributeData.ApplicationSyntaxReference.GetSyntaxAsync() as AttributeSyntax
+             ?? throw new InvalidOperationException ($"member {_attributeData} is not a AttributeSyntax");
```
Exception messages have periods at the end

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionModelBuilder.cs:
```diff
@@ -76,23 +86,19 @@ public ExpectedExceptionModel Build (AttributeData attributeData)
       }
+      
+      if(attributeData == null)
+        throw new InvalidOperationException($"Required field {nameof(attributeData)} is null.");
```
Formatting

Stefan-Ilic commented on Core/ExpectedException/IExpectedExceptionModel.cs:
```diff
@@ -24,11 +24,11 @@ namespace NUnit2To3SyntaxConverter.ExpectedException
-    ExpressionSyntax UserMessage { get; }
-    ExpressionSyntax MatchType { get; }
-    ExpressionSyntax ExpectedMessage { get; }
+    ExpressionSyntax? UserMessage { get; }
```
It's showing that still only the implementation's getters are being accessed. This could be an indication of a smell, have you double checked if this is okay?

Stefan-Ilic commented on Core/NUnitMigrator.cs:
```diff
@@ -35,36 +36,39 @@ public NUnitMigration (MigrationOptions options)
-          .Where (document => _options.SourceFileFilter.Invoke (document))
-          .Select (async document => (Original: document, New: await ConvertDocument (document, new ExpectedExceptionDocumentConverter())))
-          .Select (async document => WriteBack ((await document).Original, (await document).New));
+      var tasks = GetDocuments(solution)
```
Formatting in this file

Stefan-Ilic commented on UnitTests/ExpectedException/FeatureIntegrationTest.cs:
```diff
@@ -25,53 +25,35 @@ namespace NUnit2To3SyntaxConverter.UnitTests.ExpectedException
-      return (rewriter, method);
-    }
-
+    private const string c_testCases = "/resources/ExpectedExceptionIntegrationTestCases.cs";
```
I believe this should have some form of "file", "path", "fileName" eg `testCasesSourceFileName` or something like that, usually plural implies some sort of collection

Stefan-Ilic commented on UnitTests/ExpectedException/FeatureIntegrationTest.cs:
```diff
@@ -25,53 +25,35 @@ namespace NUnit2To3SyntaxConverter.UnitTests.ExpectedException
+      var (_, expectedSyntax) = CompiledSourceFileProvider.LoadMethod (c_expectedTestResults, method);
 
       Assert.That (
           rewriter.VisitMethodDeclaration (syntax).ToFullString().Trim(),
```
I missed this the last time, but your "Act" step is calling `VisitMethodDeclaration` meaning it should probably be in its own line

Stefan-Ilic commented on UnitTests/Helpers/CompiledSourceFileProvider.cs:
```diff
@@ -27,28 +27,45 @@ namespace NUnit2To3SyntaxConverter.UnitTests
+      return (attributeData, attributeSyntax);
+    }
+
     public static (IMethodSymbol, MethodDeclarationSyntax) LoadMethod (string file, string methodName)
```
It looks like the first tuple item in the return type is not needed anywhere

Stefan-Ilic commented on UnitTests/Helpers/CompiledSourceFileProvider.cs:
```diff
@@ -27,28 +27,45 @@ namespace NUnit2To3SyntaxConverter.UnitTests
 {
   public class CompiledSourceFileProvider
   {
+    public static (AttributeData, AttributeSyntax) CompileAttribute (string attribute)
```
The second tuple item does not seem to be necessary seeing as no test is using it

Stefan-Ilic commented on UnitTests/Helpers/CompiledSourceFileProvider.cs:
```diff
@@ -27,28 +27,45 @@ namespace NUnit2To3SyntaxConverter.UnitTests
 {
   public class CompiledSourceFileProvider
```
If this is just a static helper you can make it static

Stefan-Ilic commented on UnitTests/ExpectedException/ExpectedExceptionModelTest.cs:
```diff
@@ -0,0 +1,226 @@
+  {
+    [Test]
+    [TestCase ("[ExpectedException]")]
+    public async Task CreateFromAttributeData_NoArgsAttribute (string attribute)
```
I don't think it's necessary to use TestCase attributes if there is only one test case

Stefan-Ilic commented on UnitTests/ExpectedException/ExpectedExceptionAttributeRemoverTest.cs:
```diff
@@ -27,16 +27,17 @@ namespace NUnit2To3SyntaxConverter.UnitTests.ExpectedException
 {
   public class RemoveExpectedAttributeTest
   {
+    private const string c_testCaseFile = "resources/RewriteMethodBodyTests.cs";
```
test  case file name and test file name don't match

Stefan-Ilic commented on UnitTests/ExpectedException/ExpectedExceptionAttributeRemoverTest.cs:
```diff
@@ -0,0 +1,54 @@
+                      .SelectMany (list => list.Attributes).First (attr => attr.Name.ToString() == "ExpectedException")));
+
+
+      var rewritten = new ExpectedExceptionAttributeRemover()
```
I don't think this is a big deal, especially here, but imo tests are usually cleaner if the "Act" step really only does one thing. In this case the remover object is instanced as well, which is probably a part of the Arrange

Stefan-Ilic commented on Core/MigrationOptions.cs:
```diff
@@ -23,14 +23,14 @@ namespace NUnit2To3SyntaxConverter
   public class MigrationOptions
   {
-    private static readonly Encoding DefaultEncoding = Encoding.Default;
+    private static readonly Encoding s_defaultEncoding = Encoding.Default;
```
Actually, I don't think this field is necessary. You could just directly assign to the autoproperty

Stefan-Ilic reviewed (changes requested):
I did find some minor stuff this time round. Also, I saw 2 comments from last time are still unresolved, we should probably tackle those too

dkudernatsch committed:
code review pt. 2

dkudernatsch committed:
reformat for code review

dkudernatsch committed:
changed 1 testcase to match implementation (wellknown exception removed)

dkudernatsch committed:
removed test solution (currently not used)
removed unused integration test cases

dkudernatsch committed:
fixed formatting of attribute tags

dkudernatsch committed:
added warning and tests for empty method and assert in last statement

dkudernatsch committed:
fixed testcases

dkudernatsch committed:
now correctly formats long and multiline expected messages
removed funtionality to convert non expression statements because its way too difficult to format correctly

dkudernatsch committed:
renamed testcase to match others

Stefan-Ilic commented on Core/ConversionWarning.cs:
```diff
@@ -0,0 +1,18 @@
+﻿using System;
```
This file has no license header :-(

Stefan-Ilic commented on Core/ConversionWarning.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace NUnit2To3SyntaxConverter
+{
+  public class ConversionWarning : Exception
```
I believe that `Exception` names should always end with `Exception` no matter what

Stefan-Ilic replied:
Also, I'm not sure if this exception design is ideal in the first place. AFAIK .NET exception always follow a simple 3 constructors pattern and  usually only carry a message and an inner exception (and the stacktrace of course). Are you sure this new exception type is needed? Using well-known .NET exceptions often carries a lot more meaning to .NET developers than defining own ones. Is there perhaps a pre-defined .NET exception that, combined with a good message, carries the same amount of information as this one does?

If after contemplating you still believe that this custom exception type is beneficial, you should still rename it to end with with `Exception` and make sure the standard ctors are provided. Also, does it need to have the `[Serializable]` attribute? I know resharper's `exception` snippet always adds it

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionMethodBodyTransformer.cs:
```diff
@@ -29,15 +29,26 @@ public class ExpectedExceptionMethodBodyTransformer : ISyntaxTransformer<MethodD
+      if (node.Body == null || node.Body.Statements.Count == 0)
+      {
+          throw new ConversionWarning (node.GetLocation(), node.Identifier.ToFullString(), "Unable to convert method with empty body.");
+      }
```
We usually leave out the curly braces if the body is only one line

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionModel.cs:
```diff
@@ -49,10 +44,15 @@ public class ExpectedExceptionModel : IExpectedExceptionModel
+    public ExpressionSyntax ExceptionType { get; }
+    public ExpressionSyntax? UserMessage { get; }
+    public ExpressionSyntax? MatchType { get; }
+    public ExpressionSyntax? ExpectedMessage { get; }
```
These still seem to be below the ctor. We usually like the class' object state to be visible at one glance, meaning above the ctor

Stefan-Ilic reviewed (changes requested):
Almost there! I just have a few questions about the new Exception type and a few minor discussion points before we're finished

dkudernatsch committed:
rewritten error handling

dkudernatsch committed:
added license headers

dkudernatsch committed:
corrected ctor position

dkudernatsch committed:
formatting fix

dkudernatsch committed:
formatting fix

dkudernatsch committed:
Added unittests for Validators
fixed implementation of lastsmtisexpressionstmtValidator

dkudernatsch committed:
added license header

dkudernatsch committed:
fixed class name

Stefan-Ilic commented on Core/ExpectedException/Validators/EmptyMethodBodyValidator.cs:
```diff
@@ -0,0 +1,34 @@
+
+#endregion
+
+using System.Drawing.Design;
```
Unnecessary using directives

Stefan-Ilic replied:
Also this validator (and possibly others) seems to be unformatted

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionRewriter.cs:
```diff
@@ -22,7 +22,9 @@
 using Microsoft.CodeAnalysis.Formatting;
+using NUnit2To3SyntaxConverter.ExpectedException.Validators;
 using Serilog;
+using Serilog.Core;
```
unused

Stefan-Ilic commented on Core/ExpectedException/ExpectedExceptionMethodBodyTransformer.cs:
```diff
@@ -29,26 +29,12 @@ public class ExpectedExceptionMethodBodyTransformer : ISyntaxTransformer<MethodD
   {
     public MethodDeclarationSyntax Transform (MethodDeclarationSyntax node, IExpectedExceptionModel model)
     {
-      if (node.Body == null || node.Body.Statements.Count == 0)
```
I believe we have reached a good compromise between readability and usage safety, for our use cases it should more than do the trick

Stefan-Ilic reviewed (approved):
I may have gone a little overboard with my comments, but I believe this was a good learning experience for us both. I did leave some formatting/minor code quality comments, so please make sure to correct them before merging. Overall a very nice job!

dkudernatsch committed:
removed unused imports

dkudernatsch committed:
formatting

dkudernatsch committed:
removed unused code
